### PR TITLE
Use the ScalaObjectMapper trait to create Jackson's ObjectMapper.

### PR DESCRIPTION
**Pull request by ghost**
When deserializing a collection type containing objects, Jackson's default `<T> T readvalue(String content, Class<T> ValueType)` method does not work properly. Jackson provided `ScalaObjectMapper` trait to solve this problem, using the `def readValue[t: manifest](content: String): T` method.

So we can reuse the same ObjectMapper object outside of akka serialization. For example, when providing JSON serialization service for HTTP.

```scala
  class ScalaObjectMapper(mapper: ObjectMapper)
      extends ObjectMapper(mapper: ObjectMapper)
        with com.fasterxml.jackson.module.scala.ScalaObjectMapper
  private val objectMapper = 
    new ScalaObjectMapper(JacksonObjectMapperProvider(system).getOrCreate("jackson-json", None))

  /**
   * HTTP entity => `A`
   */
  implicit def unmarshaller[A: Manifest]: FromEntityUnmarshaller[A] =
    Unmarshaller.byteStringUnmarshaller.forContentTypes(MediaTypes.`application/json`).mapWithCharset {
      case (ByteString.empty, _) => throw Unmarshaller.NoContentException
      case (data, charset)       => objectMapper.readValue[A](data.decodeString(charset.nioCharset.name))
    }

  /**
   * `A` => HTTP entity
   */
  implicit def marshaller[A]: ToEntityMarshaller[A] =
    Marshaller.withFixedContentType(ContentTypes.`application/json`)(obj => objectMapper.writeValueAsString(obj))

  test("ScalaObjectMapper") {
    val vector = Vector(Name("Yang", "Jing"), Name("Yang", "Xunjing"))
    val entity = HttpEntity(ContentTypes.`application/json`, objectMapper.writeValueAsString(vector))
    entity.contentType shouldBe ContentTypes.`application/json`
    Unmarshal(entity).to[Vector[Name]].futureValue shouldBe vector
  }
```

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @yangbajing,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Review by octonato — COMMENTED** [review]
Hi @yangbajing, 

There are a few issues with this PR. First, it's breaking binary-compatibility (see my comment below). 

Second, is this really needed? 

Have a look at this class and test:

https://github.com/yangbajing/akka/blob/feature/scala-object-mapper/akka-serialization-jackson/src/test/scala/akka/serialization/jackson/JacksonSerializerSpec.scala#L75

https://github.com/yangbajing/akka/blob/feature/scala-object-mapper/akka-serialization-jackson/src/test/scala/akka/serialization/jackson/JacksonSerializerSpec.scala#L732-L737

`CollectionsCommand` contains a `Vector[SimpleCommand]` and the test proves that we can serialize and deserialize it correctly. 

I guess this is happening with you because you are passing the Vector directly instead of wrapping it into some other classes. 

Maybe a better approach here is to have `ScalaObjectMapper` in your codebase and use it inside the marshallers you are defining without adding it to Akka. Or to wrap the `Vector` in some other class.

Would that work for you?

File: akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonObjectMapperProvider.scala [hunk 1]
```diff
@@ -229,7 +229,7 @@ object JacksonObjectMapperProvider extends ExtensionId[JacksonObjectMapperProvid
       objectMapperFactory: JacksonObjectMapperFactory,
       config: Config,
       dynamicAccess: DynamicAccess,
-      log: Option[LoggingAdapter]): ObjectMapper = {
+      log: Option[LoggingAdapter]): ScalaObjectMapper = {
```
[the remark this answers is not in the document]
Comment (yangbajing) [reply akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonObjectMapperProvider.scala]: Yes, maybe it's better to encapsulate it in the user code. I close this pr

[the remark this answers is not in the document]
Comment (renatocaval) [reply akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonObjectMapperProvider.scala]: Thanks @yangbajing. 

And thanks for bring that up, it made me discover a few things about Jackson serialization in Akka. ;-)

**Pull request closed without merging**